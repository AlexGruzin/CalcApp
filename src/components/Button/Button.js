import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Spinner from 'components/Spinner';
import classNames from 'classnames';

import './Button.styles.less';

/* eslint-disable react/button-has-type */
const Button = ({ title, onClick, isLoading, loadingPromise, disabled, className, children, type, value }) => {
  const [content, setContent] = useState(isLoading ? <Spinner /> : title);
  const [isDisabled, setIsDisabled] = useState(isLoading);

  useEffect(() => {
    if (loadingPromise) {
      setContent(<Spinner />);
      setIsDisabled(true);

      loadingPromise.finally(() => {
        setIsDisabled(false);
        setContent(title);
      });
    }
  }, [loadingPromise, title]);

  useEffect(() => setIsDisabled(isLoading), [isLoading]);

  useEffect(() => setContent(isLoading ? <Spinner /> : title), [title, isLoading]);

  return (
    <button
      disabled={isDisabled || disabled}
      type={type}
      value={value}
      className={classNames('custom-button', className, { isDisabled: disabled })}
      onClick={onClick}
    >
      {content || children}
    </button>
  );
};

Button.propTypes = {
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  onClick: PropTypes.func,
  className: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.string, PropTypes.array]),
  isLoading: PropTypes.bool,
  loadingPromise: PropTypes.shape({
    then: PropTypes.func.isRequired,
    catch: PropTypes.func.isRequired,
    finally: PropTypes.func.isRequired,
  }),
  disabled: PropTypes.bool,
  type: PropTypes.string,
  value: PropTypes.string,
};

Button.defaultProps = {
  title: '',
  type: 'button',
  value: '',
  isLoading: false,
  loadingPromise: null,
  disabled: false,
  className: '',
  children: null,
  onClick: () => undefined,
};

export default Button;
