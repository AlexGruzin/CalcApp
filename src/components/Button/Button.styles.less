.custom-button {
  cursor: pointer;
  width: 142px;
  height: 25px;
  border: 2px solid transparent;
  font-family: "MontserratMedium", sans-serif;
  color: var(--white);
  font-size: var(--fontSize-tiny);
  font-weight: 600;
  letter-spacing: 0;
}

.custom-button.isDisabled {
  background: var(--grey300);
}

.custom-button:focus,.custom-button:active {
  box-shadow: none;
  outline: 0;
  border: 2px solid var(--outline-default);
}

.custom-button {
  &_blue {
    background: var(--blue-dark);
  }
}
