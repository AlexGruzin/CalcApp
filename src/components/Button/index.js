import { withTranslation } from 'react-i18next';
import Button from './Button';

export default withTranslation()(Button);
