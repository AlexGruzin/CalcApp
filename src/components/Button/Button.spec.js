import React from 'react';
import { waitFor, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Button from './Button';

describe('Slider tests', () => {
  it('Check snapshots of Slider are equal.', () => {
    const { asFragment } = render(<Button title="title" onChange={() => undefined} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Slider do call onClick.', async () => {
    const mockOnClick = jest.fn();
    const { getByText } = render(<Button title="title" onClick={mockOnClick} />);

    const btn = getByText('title');
    userEvent.click(btn);

    await waitFor(() => {
      expect(mockOnClick).toBeCalledTimes(1);
    });
  });
  // it.skip('Slider renders spinner when promise provided.', () => {}); // loadingPromise
  // it.skip('Slider renders spinner when isLoading.', () => {}); // isLoading
});
