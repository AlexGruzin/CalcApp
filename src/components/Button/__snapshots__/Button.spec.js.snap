// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider tests Check snapshots of Slider are equal. 1`] = `
<DocumentFragment>
  <button
    class="custom-button"
    type="button"
    value=""
  >
    title
  </button>
</DocumentFragment>
`;
